proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1018 @@
+``StrLike`` type class::
+
+    class StrLike str where
+        null :: str
+        append :: str -> str -> str
+
+    concat :: StrLike str => [str] -> str
+    concat xs = foldr append null xs
+
+Clearly, type classes can reduce some of the encoding overhead
+seen in dictionary-passing style.  But it does not solve all
+the issues of dictionary-passing style (performance, composibility,
+and applicability to the package level all apply to type classes),
+and they introduce some issues of their own:
+
+1. If there is no natural type for what we are modularizing
+   over (e.g., we are simply parametric over an implementation),
+   type class resolution cannot be carried out without introducing
+   a dummy proxy type.
+
+2. In a multi-parameter type class, some methods may not mention
+   all of the types in the type class: these methods cannot be
+   resolved unambiguously without introducing a functional dependency.
+   Similarly, every type parameter must be listed in the constraints,
+   even if they are not used.  If there is a natural type to
+   modularize over, these issues can be alleviated by introducing
+   an *associated type*.
+
+3. And of course, you still have to put the damn constraint on
+   every parametrized function.
+
+**Preprocessor.**  A classic way to swap out a backing implementation
+is to replace it textually using a pre-processor.  In such a scheme, you
+might write your Haskell program like this::
+
+    import Data.ByteString as Str
+
+    #include "Impl.hs"
+
+Besides the intrinsic ugliness of such an approach, it fails to define
+the *interface* between an implementation and the backing
+implementations it relies on.  Without this type information, separate
+typechecking is impossible.  Separate typechecking is extremely
+useful: otherwise, you get extremely difficult to understand error
+messages, as is seen in C++ template error messages.
+
+Overview
+~~~~~~~~
+
+In this section, we give a high-level overview of Backpack, using the
+string-processing problem as our running example. In this section, we
+will introduce important terminology (**bolded**) that we will be used
+in the rest of this specification.
+
+Backpack solves this string-processing problem in the following way:
+To parametrize over a string implementation, a user of Backpack writes a
+**signature** describing the necessary supported interface for their
+strings  (an *egalitarian interface.*) A signature is much like an
+``hs-boot`` file, in that it contains only type signatures and type
+declarations, but no implementations::
+
+    -- in Str.hsig
+    signature Str where
+        data Str
+        null   :: Str
+        append :: Str -> Str -> Str
+
+Other modules in the library import this signature and can use
+the types and functions declared in the signature as if it were
+an ordinary module (i.e., *just normal Haskell*, with *no copy-pasting*)::
+
+    -- in Concat.hs
+    module Concat where
+        import Str
```

Comment:
It makes matters clearer when you have multiple modules. But it's unnecessary here, so I'll dedent it.